### PR TITLE
Security2

### PR DESCRIPTION
**Pull request by ArPimentel**
[ x ] Created user and role for the connexion with frontend and creating jwt

### CODE REVIEW

**Review by remy-rm — APPROVED** [review]

File: src/main/java/com/templateproject/api/config/KeyGenerator.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package com.templateproject.api.config;
+
+import org.springframework.stereotype.Component;
+
+import java.security.SecureRandom;
+import java.util.Base64;
+
+@Component
+public class KeyGenerator {
+    private static final int DEFAULT_KEY_LENGTH = 32; // 256 bits
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/config/KeyGenerator.java]: remove comments

File: src/main/java/com/templateproject/api/config/SecurityConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+
+
+@Configuration
+
+@EnableWebSecurity
+public class SecurityConfiguration {
+
+    //Access to secret key from application.properties, available in the whole application
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/config/SecurityConfiguration.java]: remove dead code

File: src/main/java/com/templateproject/api/config/SecurityConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    @Value("${your.jwt.secret}") // Replace with the actual property key for your jwt secret from your application.properties
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/config/SecurityConfiguration.java]: remove dead code

File: src/main/java/com/templateproject/api/entity/User.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+@Entity
+@Table(name = "users")
+public class User implements UserDetails {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "user_id")
+    private Long id;
+
+//    @NotBlank(message = "Username is mandatory")
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/entity/User.java]: remove dead code

File: src/main/java/com/templateproject/api/entity/User.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    @Email
+    @Column(unique=true, nullable=false)
+    private String email;
+
+
+    @Column( name = "firstname")
+    private String firstname;
+
+    //EAGER, will directly fetch the roles, longer loading time
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/entity/User.java]: remove dead code

File: src/main/java/com/templateproject/api/repository/UserGeneratorRepository.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        userRoles.add(userRole);
+
+        Set<Role> adminRoles = new HashSet<>();
+        adminRoles.add(userRole);
+        adminRoles.add(adminRole);
+
+        User user = new User(
+                passwordEncoder.encode("password"),
+                faker.internet().emailAddress(),
+                "user_first_name", // Replace this with an actual first name or leave it as is
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/repository/UserGeneratorRepository.java:47]: remove dead code

File: src/main/java/com/templateproject/api/repository/UserGeneratorRepository.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                userRoles
+        );
+        User admin = new User(
+                passwordEncoder.encode("password"),
+                faker.internet().emailAddress(),
+                "admin_first_name", // Replace this with an actual first name or leave it as is
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/repository/UserGeneratorRepository.java]: remove dead code

File: src/main/java/com/templateproject/api/service/TokenService.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+public class TokenService {
+
+    private final JwtEncoder jwtEncoder;
+    private final byte[] jwtSecret;
+
+    public TokenService(JwtEncoder jwtEncoder) {
+        this.jwtEncoder = jwtEncoder;
+        this.jwtSecret = Base64.getDecoder().decode("VzyVUCJujcNQi0iK8uSEVJc1MdjDkBx/E5ru+I0KoD8=");
+    }
+    //Use Authentication class from security ???
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/TokenService.java]: remove dead code

File: src/main/java/com/templateproject/api/service/TokenService.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    public String generateToken(Authentication auth) {
+        // We want to create a JWT token
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/TokenService.java]: remove dead code

File: src/main/java/com/templateproject/api/service/TokenService.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        // Choose and set the algorithm used to sign the token (here HS256) and build the header
+        JwsHeader jwsHeader = JwsHeader.with(() -> "HS256").build();
+
+        // Retrieve the timestamp of the moment the token is created
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/TokenService.java]: remove dead code

File: src/main/java/com/templateproject/api/service/TokenService.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+        Instant now = Instant.now();
+
+        // We want to set the expiration date of the token to 30 days after its creation
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/TokenService.java]: remove dead code

File: src/main/java/com/templateproject/api/service/TokenService.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        Instant expiration = now.plusSeconds(30 * 24 * 60 * 60);
+
+        // We want to set the scope of the token to the authorities of the user
+        String scope = auth.getAuthorities().stream()
+                .map(GrantedAuthority::getAuthority)
+                .collect(Collectors.joining(" "));
+
+        // We want to set the claims of the token
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/TokenService.java]: remove dead code

File: src/main/java/com/templateproject/api/service/TokenService.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        JwtClaimsSet claims = JwtClaimsSet.builder()
+                .issuer("self")
+                .issuedAt(now)
+                .expiresAt(expiration)
+                .subject(auth.getName())
+                .claim("scope", scope)
+                .claim("username", ((User) auth.getPrincipal()).getUserStringName())
+                .build();
+
+        // We want to encode the token with the header and the payload
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/TokenService.java]: remove dead code

File: src/main/java/com/templateproject/api/service/UserService.java [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        return userRepository.findAll();
+    }
+
+    public User getUser(Long id) {
+        return userRepository.findById(id).get();
+    }
+
+
+    public User register(String password, String email, String firstName) {
+        // We want to encode the password before saving it to the database
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/UserService.java]: remove dead code

File: src/main/java/com/templateproject/api/service/UserService.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        String encodedPassword = passwordEncoder.encode(password);
+        // We want to save the user with the role USER
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/UserService.java]: remove dead code

File: src/main/java/com/templateproject/api/service/UserService.java [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+        Role role = roleRepository.findByAuthority("ROLE_USER").orElseThrow(() -> new RuntimeException("ROLE_USER not found"));
+        Set<Role> roles = new HashSet<>();
+        roles.add(role);
+        // We want to create a new user
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/UserService.java]: remove dead code

File: src/main/java/com/templateproject/api/service/UserService.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+        User user = new User(encodedPassword, email, firstName, roles);
+
+        // Set the username to be the same as the email
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/UserService.java]: remove dead code

File: src/main/java/com/templateproject/api/service/UserService.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+        user.setUsername(email);
+
+        // We want to save the user to the database
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/service/UserService.java]: remove dead code

File: src/main/java/com/templateproject/api/utils/GsonFactory.java [hunk 19]
```diff
@@ -0,0 +1,15 @@
+package com.templateproject.api.utils;
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/utils/GsonFactory.java]: this file is useless and never used, remove it

File: src/main/resources/logback.xml [hunk 20]
```diff
@@ -0,0 +1,11 @@
+<configuration>
```
Comment (Benja-san) [inline src/main/resources/logback.xml]: this file is useless and never used, remove it

File: src/main/java/com/templateproject/api/utils/InstantTypeAdapter.java [hunk 21]
```diff
@@ -0,0 +1,35 @@
+package com.templateproject.api.utils;
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/utils/InstantTypeAdapter.java]: this file is useless and never used, remove it

File: pom.xml [hunk 22]
```diff
@@ -58,16 +63,41 @@
...
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-oauth2-resource-server</artifactId>
+			<version>3.1.1</version>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.security</groupId>
+			<artifactId>spring-security-oauth2-jose</artifactId>
+			<version>6.1.1</version>
+		</dependency>
+		<dependency>
```
Comment (Benja-san) [inline pom.xml]: this dependency is never used, remove it

File: pom.xml [hunk 23]
```diff
@@ -58,16 +63,41 @@
...
+			<groupId>com.github.javafaker</groupId>
+			<artifactId>javafaker</artifactId>
+			<version>1.0.2</version>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-oauth2-resource-server</artifactId>
+			<version>3.1.1</version>
+		</dependency>
+		<dependency>
```
Comment (Benja-san) [inline pom.xml]: this dependency is never used, remove it

File: src/main/java/com/templateproject/api/ApiApplication.java [hunk 24]
```diff
@@ -1,13 +1,30 @@
 package com.templateproject.api;
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/ApiApplication.java:1]: this file is useless and never used, remove it

File: src/main/java/com/templateproject/api/config/KeyGenerator.java [hunk 25]
```diff
@@ -0,0 +1,27 @@
+package com.templateproject.api.config;
```
Comment (Benja-san) [inline src/main/java/com/templateproject/api/config/KeyGenerator.java]: this file is useless and never used, remove it

**Review by Benja-san — CHANGES_REQUESTED** [review]

**Pull request merged**